### DISCUSSION TITLE
Harmonize new recursive functions with those in recursive-refac

### DISCUSSION THREAD

**Discussion by catherinebirney** [USEPA/flowsa#285] [General]
Discuss steps necessary to update the FBS methodology to use recursive modeling
@matthewlchambers @bl-young 

- Flag new functions vs existing functions
- Improve/update documentation of FBS methods

**Comment by bl-young** [comment]
- Drop new FBS methods that are not designed to be core methods (and activitysets)

**Comment by bl-young** [comment]
Status of FBS:

- [x] GRDREL_national ([no diff](https://github.com/USEPA/flowsa/actions/runs/3349812766/jobs/5550162518))
- [x] GRDREL_state ([no diff](https://github.com/USEPA/flowsa/actions/runs/3351681232/jobs/5553293491))
- [x] TRI_DMR_national ([diff](https://github.com/USEPA/flowsa/actions/runs/3357790305)) - reflects changes in how `MetaSources` are aggregated (they are not in new branch, so more records in the final FBS); no difference in quantities
- [x] TRI_DMR_state
- [x] CRHW_national ([no diff](https://github.com/USEPA/flowsa/actions/runs/3357775325/jobs/5563870212))
- [x] CRHW_state
- [ ] CAP_HAP_national (requires significant refactoring, currently causing [memory error](https://github.com/USEPA/flowsa/actions/runs/3353799708/jobs/5556877369))
- [x] CNHW_CA ([no diff](https://github.com/USEPA/flowsa/actions/runs/3382381414/jobs/5617246059))
- [x] CNHW_national ([no diff](https://github.com/USEPA/flowsa/actions/runs/3379422031/jobs/5610952527))
- [x] CNHW_state ([no diff](https://github.com/USEPA/flowsa/actions/runs/3379423478/jobs/5610956297))
- [x] CNHWC_national ([no diff](https://github.com/USEPA/flowsa/actions/runs/3382454893/jobs/5617398077))

**Reply by bl-young** [reply]
Both TRI_DMR and CRHW seem to show unaggregated records that if aggregated would appear the same as remote. I thought https://github.com/USEPA/flowsa/commit/acc7aabeb6fad1113af69830ef4e20344879c29a would do it, but still showing multiple records for individual flows and sectors that together sum to the expected total.

**Reply by bl-young** [reply]
Ok I adjusted where the aggregation takes place in https://github.com/USEPA/flowsa/commit/a685318d52a881e1ef45b53e87701b2b30d8b05a and that solves the issues with CRHW and TRI_DMR

**Reply by matthewlchambers** [reply]
What do the checkmarks refer to here? Just wondering since the `GRDREL_national`, at least, doesn't run using the new code.

**Reply by bl-young** [reply]
@matthewlchambers Yes I did link to the actions confirming being able to run them. I'm surprised you're not able to run GRDREL_national...

On recursive-refac I can succesfully run 
```
    m = 'GRDREL_national_2017'
    flowsa.flowby.FlowBySector.generateFlowBySector(m)
```

**Reply by bl-young** [reply]
But yes the checkmarks reflect successful completion of the method update (also trying to note whether the results differ or not, and perhaps why)

**Reply by matthewlchambers** [reply]
Ah, I see. That runs the old code, not the new code.

EDIT: We determined that this assertion is false.

**Reply by bl-young** [reply]
> Ah, I see. That runs the old code, not the new code.

I'm not sure I follow. I'm able to run `GRDREL_national` with your new updates with `generateFlowBySector()`. What do you mean by the new code?

**Comment by matthewlchambers** [comment]
I didn't modify `flowbysector.py`, so the new code for generating an FBS is accessed via `flowsa.flowby.FlowBySector.generateFlowBySector()`, not `flowsa.flowbysector.generateFlowBySector()`.

**Reply by bl-young** [reply]
Right- the code above uses flowby

**Reply by matthewlchambers** [reply]
Well I feel silly, haha. I misread it somehow. My apologies!

**Reply by bl-young** [reply]
Haha no problem! But does that still mean you are unable to run it?

**Reply by matthewlchambers** [reply]
Solved the issue. I just needed to delete some of the saved `stewi` parquets (my local copy was from last year or something like that). Once I did that everything worked fine!

**Comment by bl-young** [comment]
@matthewlchambers I'm looking for some help on `CNHW_CA_2014` which yields a key error in convert_to_geoscale after my [method updates](https://github.com/USEPA/flowsa/commit/4a109f094e7e41e5f881a406e608cc0ad9cb86fd)
```
 File ~\Desktop\Spyder\flowsa\src\flowsa\flowsa\flowby.py:1647 in generateFlowBySector
    fbs = pd.concat([

  File ~\Desktop\Spyder\flowsa\src\flowsa\flowsa\flowby.py:1648 in <listcomp>
    get_flowby_from_config(

  File ~\Desktop\Spyder\flowsa\src\flowsa\flowsa\flowby.py:1369 in prepare_fbs
    pd.concat([

  File ~\Desktop\Spyder\flowsa\src\flowsa\flowsa\flowby.py:1370 in <listcomp>
    fba.prepare_fbs()

  File ~\Desktop\Spyder\flowsa\src\flowsa\flowsa\flowby.py:1382 in prepare_fbs
    self

  File ~\Desktop\Spyder\flowsa\src\flowsa\flowsa\flowby.py:956 in convert_to_geoscale
    self.full_name.split('.')[-1], self.config['names']

KeyError: 'names'
```

**Reply by matthewlchambers** [reply]
Just pushed a fix for this issue.

**Comment by bl-young** [comment]
As well on `CNHW_national_2014` which errors at:
```
File ~\Desktop\Spyder\flowsa\src\flowsa\flowsa\flowby.py:1647 in generateFlowBySector
    fbs = pd.concat([

  File ~\Desktop\Spyder\flowsa\src\flowsa\flowsa\flowby.py:1648 in <listcomp>
    get_flowby_from_config(

  File ~\Desktop\Spyder\flowsa\src\flowsa\flowsa\flowby.py:54 in get_flowby_from_config
    return FlowBySector.getFlowBySector(

  File ~\Desktop\Spyder\flowsa\src\flowsa\flowsa\flowby.py:1589 in getFlowBySector
    return super()._getFlowBy(

TypeError: _getFlowBy() got multiple values for keyword argument 'full_name'
```
Note this method uses an FBS as the primary data source, which is unsually and I expect is causing this issue (seems to be an issue in the recursion). @matthewlchambers

**Reply by matthewlchambers** [reply]
Just pushed a fix for this issue.

**Reply by bl-young** [reply]
thank you! I will try out both

**Comment by matthewlchambers** [comment]
Regarding `CAP_HAP_national_2017`, is there a need for using QCEWs from different years? It would be much simpler to just use a single year.

**Reply by bl-young** [reply]
Its an old helper source, so its used to expand EIA_MECS (2014 or 2018) to 6-digit NAICS, and also used to expand BEA data (2012) down to 6-digit NAICS. Are you attending today's flowsa meeting? We could put it on the agenda to discuss then.

Also there has been discussion of moving to an employment FBS instead of BLS_QCEW (aka an already processed version) that would speed things up substantially and make caching less necessary. @catherinebirney

**Reply by catherinebirney** [reply]
I think we can go ahead and use the existing Employment FBS because those employment FBS are created using the same steps we use to clean up BLS data every time it is called. We discussed filling in some of the suppressed QCEW data with QWI data, which we haven't done yet. 

The issue is that with the original method, it really wasn't set up to use FBS as an allocation source for FBAs. So depending on how Matthew's branch is set up, we could go ahead and make that switch. Although we need to update the employment FBS to use Matthew's revised method to estimate suppressed data.

**Reply by matthewlchambers** [reply]
There shouldn't be any issue with using an FBS as an allocation source in the `recursive-refac` branch, and it would be an especially good idea for anything below the national geoscale, since it takes a **looong** time to clean the QCEW at a state or county level.

**Reply by bl-young** [reply]
> since it takes a **looong** time to clean the QCEW at a state or county level.

I'm two hours in... https://github.com/USEPA/flowsa/actions/runs/3378018922/jobs/5607673435

**Comment by catherinebirney** [comment]
@matthewlchambers @bl-young 
In the recursive version, when an FBA is mapped to the fed elem flow list [here](https://github.com/USEPA/flowsa/blob/recursive-refac/flowsa/flowby.py#L1393), the FBA columns "FlowName" and "Compartment" are dropped. The next step is to use the "clean_fba" function. 

In the original method, those columns are [retained ](https://github.com/USEPA/flowsa/blob/recursive-refac/flowsa/flowbysector.py#L179). 

In the water FBS method I'm running into an issue with the recursive method that the "clean_fba" functions subset data using FlowName/Compartment. I can update the functions to use Flowable/Context info, but before I do I wanted to check if this is a widespread issue and if it is worth retaining FlowName/Compartment and dropping those columns later.

**Reply by bl-young** [reply]
So this specifically is for methods where the flows ARE mapped to FEDEFL, but we also want to maintain the original names for some reason? I am trying to remember if that is used in other methods or not...

**Reply by catherinebirney** [reply]
Correct - and I believe the original columns are maintained just for use in the "clean_fba" function and then not called on again. I'll do a little digging, but will likely update the water clean_fba to use Flowable/Context. This issue is the reason GH actions are failing.

**Reply by bl-young** [reply]
noting that approach for mapping and standardizing was revised https://github.com/USEPA/flowsa/commit/651b62f9e59c51c1ee71251fcf3dfb2110e54b1a

**Comment by catherinebirney** [comment]
FBS status:

- [x]  Employment_national_2017
- [x]  Employment_national_2018
- [x] Employment_state_2012
- [x] Employment_state_2013
- [x] Employment_state_2014
- [x] Employment_state_2015
- [x] Employment_state_2016
- [x] Employment_state_2017
- [x] Employment_state_2018
- [x] Employment_state_2019
- [x] Employment_state_2020
- [ ] Land_national_2012
- [ ] Water_national_2010_m1
- [ ] Water_national_2010_m2
- [ ] Water_national_2015_m1
- [ ] Water_national_2015_m2
- [ ] Water_national_2015_m3
- [ ] Water_state_2015_m1

**Reply by catherinebirney** [reply]
The national employment FBS run locally, but run into a memory failure on GH actions

**Comment by bl-young** [comment]
Potential feature:
Could we cache a partially processed dataset but still allow for additional `selection_fields` after its pulled from the cache?

The specific application I'm thinking is when we use EIA_MECS but need to subset for different flows. Estimating the suppressed data would only then be required once, instead of each time that attribution source is called.

https://github.com/USEPA/flowsa/blob/651b62f9e59c51c1ee71251fcf3dfb2110e54b1a/flowsa/flowby.py#L1005-L1008

I think that here we could just append `select_by_fields()` once we ensure the configuration is set up appropriately (would need to change how we check if a dataset is cached since the `isinstance(fba.config['attribution_source'], str)` would not longer apply)

**Reply by bl-young** [reply]
see #294

**Reply by matthewlchambers** [reply]
#294 has now been merged, so we now treat the cached source as any other FlowBySector (and note that even if the cached source was originally an FBA, it will have been converted to an FBS before being cached), so we can select by fields or apply cleaning functions or anything else we could do with an FBS attribution source.

**Reply by bl-young** [reply]
Yes, I noticed that I had to change the select_fields for flows to be `Flowable` instead of `FlowName` for instance in my testing. I plan to push a revised CAP_HAP soon

**Comment by bl-young** [comment]
FlowBySectors seem to generate with extra FBA columns such as: 
```
fbs.columns
Out[12]: 
Index(['Flowable', 'Class', 'SectorProducedBy', 'SectorConsumedBy',
       'SectorSourceName', 'Context', 'Location', 'LocationSystem',
       'FlowAmount', 'Unit', 'FlowType', 'Year', 'MeasureofSpread', 'Spread',
       'DistributionType', 'Min', 'Max', 'DataReliability',
       'TemporalCorrelation', 'GeographicalCorrelation',
       'TechnologicalCorrelation', 'DataCollection', 'MetaSources', 'FlowUUID',
       'ActivityProducedBy', 'ActivityConsumedBy', 'SourceName', 'Description',
       'ProducedBySectorType', 'ConsumedBySectorType'],
      dtype='object')
```
specifically: ActivityProducedBy, ActivityConsumedBy, SourceName, Description, but all fields are blank
is that by design @matthewlchambers

**Reply by bl-young** [reply]
this seems specific to `CNHW_CA_2014`

**Reply by matthewlchambers** [reply]
It's definitely not intentional. And it appears to be a more general issue (at the very least, it's also happening with `GHGI_recursive_2017_m1`).

**Reply by matthewlchambers** [reply]
I'll have to figure out what's happening here. Thanks for pointing it out!

**Reply by matthewlchambers** [reply]
Ok, see #295 for a fix.

**Reply by matthewlchambers** [reply]
Merged #295 into `recursive-refac`.

**Comment by catherinebirney** [comment]
@matthewlchambers @bl-young The current hold up with converting Water_national_m1 to the recursive method is that the 'clean_fba' function is called on for each activity set individually, rather than applying 'clean_fba' to the entire source df before subsetting the df into activity sets (which is the order in the original method). For USGS_NWIS_WU, the df needs to be cleaned before broken into activity sets because there are some calculations that require activities that are in different activity sets. 

@matthewlchambers Is there a reason 'clean_fba' is called on repeatedly rather than a single time? Can we call on clean_fba before each activity set?

**Reply by bl-young** [reply]
would this work: [clean_fba_before_activity_sets](https://github.com/USEPA/flowsa/blob/recursive-refac/flowsa/flowby.py#L1406)

**Reply by catherinebirney** [reply]
@bl-young - yes, thank you!

**Comment by matthewlchambers** [comment]
Just pushed an update to `recursive-refac` changing how primary and secondary columns are assigned when we specify `primary_action_type`. Specifically, I realized that the primary column was being assigned in an efficient way, but the secondary column was being assigned inefficiently. Now, both columns are assigned efficiently if `primary_action_type` is specified, and it takes 10-15 seconds to create each activity set in `CAP_HAP_national_2017`. Let me know if you encounter any issues.

**Reply by bl-young** [reply]
oh nice - thanks! i'll check it out

**Reply by bl-young** [reply]
this is making my life so much easier 🙌

**Comment by bl-young** [comment]
The crop data in NEI via USDA uses `disaggregate_coa_cropland_to_6_digit_naics`. @matthewlchambers  is that one that you've converted yet? I also see for GHGI you are using a second attribution of `USDA_CoA_Cropland_NAICS`, is that the best model to follow?

**Reply by matthewlchambers** [reply]
What `disaggregate_coa_cropland_to_6_digit_naics` seems to be doing is using `USDA_CoA_Cropland_NAICS` as an attribution source. So yes, I think just explicitly using `USDA_CoA_Cropland_NAICS` as a secondary attribution source is better if possible. If there's something more going on with `disaggregate_coa_cropland_to_6_digit_naics` than I realized, it can probably still be implemented in the context of `USDA_CoA_Cropland_NAICS` as a secondary attribution source.

**Reply by matthewlchambers** [reply]
For example, as a `clean_fba` function for `USDA_CoA_Cropland_NAICS`.

**Comment by bl-young** [comment]
@matthewlchambers I'm experiencing an issue with CAP_HAP when trying to use BEA Use table data to allocate an activity set (e.g. `EPA_NEI_Nonpoint.mobile_and_fuel_storage`). Perhaps this was not relevant for the GHG data because both the primary source and attribution source are the same geoscale.

https://github.com/USEPA/flowsa/blob/eb3724ab04503719de5e49249d11e4b338c62ec0/flowsa/flowby.py#L901-L917

When creating the `highest_reporting_level_by_geoscale` from BEA, one of the Activity columns is typically `""` due to the collapse of the columns:
https://github.com/USEPA/flowsa/blob/eb3724ab04503719de5e49249d11e4b338c62ec0/flowsa/methods/flowbysectormethods/CAP_HAP_national_2017.yaml#L160-L161

In line 915 in the chunk above, converting to `.astype(object)` seems to convert those `""` into `nan`. Then upon subsqueently merging with the FBA, the merge fails due to those `nan`. 

Notably adding a `.fillna('')` solves this problem, but I expect that is not a good fix. Is `.astype(object)` necessary here?

**Reply by catherinebirney** [reply]
@bl-young the use of .astype(object) is a recent addition I made to get the water FBS to work. When creating the two dataframe before the merge, one dataframe ActivityProducedBy column is all np.nan, which becomes a float64 column. The other df has a mix of string and np.nan values in the ActivityProducedBy column and is an object column. The merge fails when the datatypes are different.

**Reply by bl-young** [reply]
Interesting. I wonder if we could ensure at somepoint that all APB and ACB values are strings and fill any nan with '' so we are consistent here.

**Reply by matthewlchambers** [reply]
@bl-young That is generally what happens, so I think what we want to do is find a different solution to the issue @catherinebirney was having with water. This code is definitely less well tested than other code because of the GHGI primary and attribution sources using the same geoscale (as you pointed out, Ben).

**Reply by matthewlchambers** [reply]
@catherinebirney when you say "before the merge", do you mean the merge in line 903?

**Reply by catherinebirney** [reply]
no, the merge on line 919. The issue occurs in Water_national_m1_2015 for the first activity set, direct_attribution

**Reply by matthewlchambers** [reply]
Just pushed a tweak to @catherinebirney 's solution, so that we explicitly set the datatypes of the `ActivityProducedBy` and `ActivityConsumedBy` columns to `object`. I'm hoping that fixes both issues.

**Reply by bl-young** [reply]
Its been a while since I've looked at the water method, though i recall there were challenges with dealing with data in both ACB and APB. I wonder if the approach similar to what's implemented with BEA data would simplify things - and allow you to do an activity set only for activities that show up in a specific column.

**Reply by bl-young** [reply]
> Just pushed a tweak to @catherinebirney 's solution, so that we explicitly set the datatypes of the `ActivityProducedBy` and `ActivityConsumedBy` columns to `object`. I'm hoping that fixes both issues.

no, still dropping out for BEA as attribution source

**Reply by bl-young** [reply]
need to confirm if this is still an issue.

**Reply by catherinebirney** [reply]
I'm back to having an issue here for the national water FBS because some of the data has info in both the FBA and FBS columns while other data only has info in one column

**Reply by catherinebirney** [reply]
The [solution](https://github.com/USEPA/flowsa/commit/4bec98b498288eda3a61310f22d41b2db9dc258d) I came up with is to convert the activity column dtypes only if the dtype is float. Even though the activity columns are set to object when the df is subset by geoscale, one of the water FBS activity col dtypes is reset to float because the column is subset from a mix of string/np.nan to only np.nan. So I first check if the col dype is float and if it is convert it. I don't get an error when running CAP_HAP, but CAP_HAP data is coming into the geoscale function with np.nan instead of '', so I think there is an upstream issue.

**Reply by catherinebirney** [reply]
@bl-young @matthewlchambers see if CAP HAP is still a problem

**Reply by bl-young** [reply]
I now get a memory error when I try to run the whole thing. But I am able to get through successfully these activity sets without error

**Comment by bl-young** [comment]
I'm having an issue in `convert_to_geoscale()` for stewi data where the dataset (e.g. RCRAInfo in `CRHW_national_2017`) comes in at both state and county level. Unlike in some other datasets (e.g. BLS data) where both state and county are reported separately (and generally these are considered duplicative), in some stewi datasets we do not have full location information for a facility and can only assign a state but not county. Therefore a full complete non-duplicative dataset will have both state and county records.

Specifically data are getting dropped here because the `geoscale == source_geoscale` line 
https://github.com/USEPA/flowsa/blob/0d0e36134c86077c47919d9ebc076d492947e7c5/flowsa/flowby.py#L934-L944

any ideas @matthewlchambers

**Reply by bl-young** [reply]
I should note that this stems from an effort to update stewi based datasets to use flowby functions and the new classes

**Reply by bl-young** [reply]
since this likely applies only to point source datasets (e.g. stewi), handle this in stewi by converting to the appropriate geoscale all entries within that function)

**Reply by bl-young** [reply]
This has been fixed here: https://github.com/USEPA/flowsa/commit/3fcc451c122092144bc40189c86c5dff1e4cd99c; handled within stewi_fbs.py

**Comment by bl-young** [comment]
Duplicate FIPS appearing in FIPS_geoscale `51019`
![image](https://user-images.githubusercontent.com/44471635/201708095-f71dae6f-15bd-4282-a786-14e93a0d38cd.png)

I expect this stems from a FIPS getting split across the time series but have not yet confirmed. I also have not yet confirmed if this is an issue in the main branch.

**Reply by matthewlchambers** [reply]
What years are involved here? Bedford (city) switched from being an independent city (FIPS code 51515; not included in any county) to being part of Bedford county (FIPS code 51019) in 2013. Virginia's practice of making incorporated cities explicitly _not_ part of any county makes any sort of time series county-level data interesting, to say the least.

**Reply by bl-young** [reply]
Ahh yes as a Virginia resident this is not the first time i've heard about Virginia causing real problems for data science! :) 
I believe this was the default (so 2015?) and I think it's due to a merge across years at some point.

**Reply by catherinebirney** [reply]
We do have a [FIPS crosswalk](https://github.com/USEPA/flowsa/blob/master/flowsa/data/FIPS_Crosswalk.csv) which could be used to align the years

**Reply by bl-young** [reply]
Yes I think that's the issue Bedford shows up twice in the crosswalk (line 2879 and 2966) because of the two fips merging

**Reply by matthewlchambers** [reply]
That could indeed do it. We may need to revisit how the crosswalk is applied and make sure we're getting our FIPS codes from the correct year.

**Comment by catherinebirney** [comment]
@matthewlchambers @bl-young 
The original FBS methods were set up to have a [target geoscale](https://github.com/USEPA/flowsa/blob/master/flowsa/methods/flowbysectormethods/Water_national_2015_m1.yaml#L7) and then within a data source, we could specify which [geoscale to use](https://github.com/USEPA/flowsa/blob/master/flowsa/methods/flowbysectormethods/Water_national_2015_m1.yaml#L12). This option is necessary for cases like the water method where the objective is a national level dataset, but we want to use state level data to get there rather than published national data. Now if within a data source we specify a geoscale (see [water m1](https://github.com/USEPA/flowsa/blob/recursive-refac/flowsa/methods/flowbysectormethods/Water_national_2015_m1.yaml#L12)), that specified geoscale overrides the target geoscale listed at the top of the yaml. I plan to modify the code so the top level geoscale is still the target - I don't think this should cause issues with other FBS methods and Ben, I believe you need this option for the CAP_HAP method

**Reply by matthewlchambers** [reply]
My only concern here is that I don't want to make `geoscale` an exception to the general recursive nature of our config files. Consequently, I wonder if there's a way to instead filter the specific datasets (like water) where we want to use state level data even though the objective is national.

**Reply by matthewlchambers** [reply]
If the water dataset has a column for geoscale, we could just do `select_by_fields()` to filter it to state level data. If not, a `clean_fba_before_mapping` function could also work, it seems.

(Caveat, of course, that I'm not familiar with the data source in question. But I do want to preserve the recursive nature of the config files if at all possible, so if these suggestions don't work, maybe we can brainstorm an alternative solution.)

**Reply by bl-young** [reply]
There was some other case where I was thinking about filtering by states (or maybe it was this case...). I think a common fxn might be better than `select_by_fields` because we'd have to write out all 50+ FIPS codes.

**Reply by bl-young** [reply]
consensus is that this is an issue where the target geoscale of the method is different (less granular) than the attribution source. Likely need a more complex handling of this.

E.g. national water method that needs to use state level data for allocation.

**Reply by matthewlchambers** [reply]
Want to confirm something: Is the primary dataset (the one being attributed) at state or national level in this case (for the relevant activity sets)?

**Reply by bl-young** [reply]
state level

**Reply by matthewlchambers** [reply]
Ok, then my first thought is that we can specify state geoscale for _both_ the activity set and the attribution source (illustration below). This will result in an FBS with different geoscales for different activity sets, but we already needed to aggregate the different activity sets up anyway.

```
# Top of document. Keys irrelevant to the current discussion omitted
geoscale: national
source_names:
  Arbitrary_Source:
    activity_sets:
      arbitrary_activity_set:
        geoscale: state
        attribution_method: proportional
        attribution_source:
          arbitrary_attribution_source:
            geoscale: state
```

**Reply by catherinebirney** [reply]
Do you mean the final FBS output for a "national" level FBS will have a mix of national and state data (for example) based on if we specify a geoscale in an activity set? If I'm understanding your comment correctly, we don't want to do that. For some of the FBS methods we want to create both national and state level versions. If the "national" version has some state-level data, there is a chance it won't align perfectly with the "state" FBS version and that would be confusing. And the objective is to have a national level FBS. Can we add an additional line of code at the end of the FBS build to get the top level geoscale and aggregate up any activity sets that do not meet it? I

**Reply by matthewlchambers** [reply]
I should also point out that the recursive nature of the configuration file means that specifying `geoscale: state` is actually unnecessary for the attribution source if it is specified on the activity set. So the example above should more properly read:

 ```
# Top of document. Keys irrelevant to the current discussion omitted
geoscale: national
source_names:
  Arbitrary_Source:
    activity_sets:
      arbitrary_activity_set:
        geoscale: state
        attribution_method: proportional
        attribution_source:
          arbitrary_attribution_source:
```

**Reply by matthewlchambers** [reply]
> Do you mean the final FBS output for a "national" level FBS will have a mix of national and state data (for example) based on if we specify a geoscale in an activity set? If I'm understanding your comment correctly, we don't want to do that. For some of the FBS methods we want to create both national and state level versions. If the "national" version has some state-level data, there is a chance it won't align perfectly with the "state" FBS version and that would be confusing. And the objective is to have a national level FBS. Can we add an additional line of code at the end of the FBS build to get the top level geoscale and aggregate up any activity sets that do not meet it? I

I think that should be very doable. In fact, it might be the case that the existing FB `convert_fips_to_geoscale()` method, combined with `aggregate_flowby()` would obtain something close to the desired result (if provided with the correct arguments, including the set of column to group by in the case of `aggregate-flowby()`). However, I think the preferred approach would be to implement an FBS version of the `convert_to_geoscale` method, currently only implemented for FBA.

**Reply by catherinebirney** [reply]
added method of aggregating activity geoscale to target geoscale with [70d4125](https://github.com/USEPA/flowsa/commit/70d41251e7f1936eacb3ab8c288bbf3d5edf4bd4)

**Comment by catherinebirney** [comment]
@matthewlchambers I have another Water FBS specific issue. The target sector is 6-digits, however some of the data are brought in at 7-digit sectors and I want to keep them at NAICS7 until after 2 datasets are [multiplied](https://github.com/USEPA/flowsa/blob/recursive-refac/flowsa/methods/flowbysectormethods/Water_national_2015_m1.yaml#L53). I am multiplying a water application rate (kg/m2) by total land area (m2) for specific crops. Aggregating to NAICS6 before multiplying would result in less accurate calculations of water use by crop type. Additionally because one of the data sets is a rate (kg/m2), that dataset cannot be aggregated from NAICS7 to NAICS6 by simply summing the data, it would need to be a weighted aggregation.

**Reply by matthewlchambers** [reply]
I think this can be approached in the same way as above, by specifying a different `industry_spec` for the activity set (which will also be applied to the attribution source), then aggregating afterward.

**Reply by catherinebirney** [reply]
I was thinking the same. There is still an issue of aggregating "rates' which right now I think is only in the water FBS. It might be worth adding a unit check and if the unit appears to be a rate with "/" we add a warning.

**Reply by catherinebirney** [reply]
added method of aggregating activity sets to target sector level with [0271c76](https://github.com/USEPA/flowsa/commit/0271c763395bf8f6d2f10d53b5ddfc0df4d4a6c2)

**Comment by catherinebirney** [comment]
@matthewlchambers A dataset with text based activities that could benefit from estimating suppressed data is in Water_national_2015_m1, "cropland_attribution", the ["USDA_CoA_Cropland"](https://github.com/USEPA/flowsa/blob/sector_mapping/flowsa/methods/flowbysectormethods/Water_national_2015_m1.yaml#L43-L58) FBA used for attribution. 

Right now, all suppressed data in USDA_CoA_Cropland is dropped in `convert_to_geoscale()` because that function is called before `attribute_flows_to_sectors()` and we aggregate the FBA in `convert_to_geoscale()`. What do you think of reorganizing those two function calls in [`prepare_fbs()`](https://github.com/USEPA/flowsa/blob/sector_mapping/flowsa/flowby.py#L1546-L1557)?

**Reply by matthewlchambers** [reply]
I think that shouldn't be necessary because no matter what our solution is, the suppressed data should be estimated before either of those method calls happen.

**Reply by matthewlchambers** [reply]
Unless I'm again wildly misunderstanding the issue, which is always possible, haha.

**Reply by catherinebirney** [reply]
the activities are mapped to sectors within `attribute_flows_to_sectors()` and the activities need to be assigned sectors to estimate suppressed data

**Reply by matthewlchambers** [reply]
I see what you're saying. Yes, I agree that suppressed data should be estimated (which as you've noted requires mapping to sectors) before we convert the geoscale. Not least because some methods of estimating suppressed data might use data from another geoscale to do so.

**Reply by matthewlchambers** [reply]
Starting to work on this issue. While we're renaming `aggregated` to `parent-child` and `disaggregated` to `flat`, could we also consider renaming `sector-like_activites` to `naics-like_activities`? This would be both clearer and more accurate to the present state of things.

Actually, what might be even better would be to use the existing `activity_schema` key. Anytime we would have checked whether `sector-like_activities` is `True`, we could instead check whether `activity_schema` starts with `'NAICS'`. I kind of favor this approach because even if we later add other activities schemas which are sector-like, we probably can't use the existing code for "sector-like activities", since the existing code in these cases is entirely NAICS focused.

**Reply by catherinebirney** [reply]
I like the suggestion to use `activity_schema` and move away from `sector-like_activities`. Those are duplicative parameters in a sense. Could we use a dictionary to assign schema by data year? We already have some NAICS 2017 schemas, but only for certain years. It would help if we could say EIA data from 2010-2016 uses NAICS_2012 while data from 2017+ uses NAICS_2017. Just an example - I'd have to look at the data again.

As an aside, we updated `aggregated` to `flat` and `disaggregated` to`parent-child` - goes to show how confusing those labels were :)

**Reply by matthewlchambers** [reply]
I like the idea of using a dictionary to assign schema by data-year, though there may be some implementation challenges.

**Reply by matthewlchambers** [reply]
Another question related to fixing this overall issue: could we switch to requiring crosswalks to actually specify their `SectorSourceName`? (Generally this amounts to specifying the relevant NAICS year.) This would clean up some code that currently tries to handle situations where that is missing. I think just giving an appropriate error so the user can fix their crosswalk would be better.

**Comment by catherinebirney** [comment]
@matthewlchambers in the water FBS, I want to attribute a single activity ["Irrigation Crop"](https://github.com/USEPA/flowsa/blob/sector_mapping/flowsa/methods/flowbysectormethods/Water_national_2015_m1.yaml#L36) in the primary data source to NAICS 6. To do so, I call on [USDA_CoA_Cropland](https://github.com/USEPA/flowsa/blob/sector_mapping/flowsa/methods/flowbysectormethods/Water_national_2015_m1.yaml#L42). The USDA_CoA_Cropland needs to be first further attributed using [USDA_CoA_Cropland_NAICS](https://github.com/USEPA/flowsa/blob/sector_mapping/flowsa/methods/flowbysectormethods/Water_national_2015_m1.yaml#L60). But what I would like to do is attribute all USDA_CoA_Cropland sectors that begin with '112' with UsDA_CoA_Cropland_NAICS that have a "FARM OPERATIONS" flowclass and all USDA_CoA_Cropland sectors that begin with '111' with USDA_CoA_Cropland_NAICS that have a "AG LAND, CROPLAND, HARVESTED" flowclass. The FBS methodology is not currently set up to attribute part of an attribution source via one method and another part via another method, correct? I can't break the primary data source into two activity sets, so really I need a way of breaking the attribution source into 2 activity sets.

**Reply by matthewlchambers** [reply]
You can definitely split an attribution source into activity sets :-) It's one of the nice features of the recursive system.
